Filter editors

lukasolson opened the pull request:
This PR adds more helpful editors to filters. Instead of a simple JSON editor, you get drop-downs with autocomplete based on your actual data.

Summary: We've added more helpful input controls for editing filters. When you edit a filter, you'll get drop-downs and text boxes with suggestions from your data instead of just a JSON editor. We've also added the ability to add custom filters without having to click on a visualization.

![jun-22-2017 10-33-24](https://user-images.githubusercontent.com/1178348/27447577-55c8e3be-5736-11e7-8e15-5654bfa65a11.gif)

Closes https://github.com/elastic/kibana/issues/5249.

Bargs commented:
Looks like a good start, I ran into some bugs though. I wasn't able to create a filter that matches a value, the console was spitting out the following error:

```
TypeError: Cannot read property 'value' of undefined
    at controller.save (filter_editor.js:98)
```

It would be cool if the field input was focused as soon as you click "Add a filter" so that the entire filter creation process can be keyboard driven. In fact, that thought led me to another one. What if instead of three individual dropdowns the form was a single input, a mini version of dave's [original mockup](https://cloud.githubusercontent.com/assets/324519/24534711/8663e5c6-1584-11e7-987e-bbaaf4fa3308.png)? It might not even require a modal, you could possibly edit the filter directly inline. It would begin to teach users how to use the query language we ultimately want to replace the lucene query syntax with. It would also give us a way to experiment with syntax for different filter types in a low risk way (because the saved searches would still contain the filter objects, not the query strings, a change to the syntax wouldn't break backwards compatibility). And last of all the parsing and autocomplete code would all be reusable in kuery. 

Maybe it's a silly idea and there are issues I'm not thinking of, but it just hit me and I got excited.

Bargs commented:
@lukasolson and I chatted about the above idea some more yesterday afternoon and we're thinking it's not precluded by the current approach in this PR. If we eventually want to make the filter editor a single input, these drop downs are a less complicated step in that direction.

Just to elaborate more on what I was talking about, I threw together the below mockup. Forgive the quality, I basically just hacked together pieces from this PR and Dave's mockup.

![filterflow](https://cloud.githubusercontent.com/assets/6239176/25503448/7c06fefc-2b68-11e7-95fb-60bdd2981799.jpg)

As I mentioned before, the modal might not be necessary at all in this scenario. Possibly the filter pill itself could be editable. On existing filters the field, operator, and value would be clickable, similar to http://documentcloud.github.io/visualsearch/

lukasolson commented:
@Bargs Could you see if you still get the error when trying to create an `is` filter?

Bargs commented:
@lukasolson seems to be working now 👍 

I noticed when switching between `is` and `is not` (and any other operator) the value gets reset. Do you think it should remain the same? I think it might be common to want to change just the operator.

Bargs commented:
I played with the editor some across Discover, Visualize, and Dashboard. Haven't looked at the code at all. Here's what I found:

* Would be nice if the escape key closed the editor, also when clicking outside the box
* Autocomplete for values works for keyword fields but not analyzed fields
* You can edit the query DSL when editing a filter, but not when creating one, which seems odd
* The ace editor sometimes takes a long time to load when you click “Edit the query DSL”
* Previous "is one of" values don’t get cleared out when changing the field
![screen shot 2017-05-11 at 11 29 24 am](https://cloud.githubusercontent.com/assets/6239176/25961532/9cc6d5dc-3648-11e7-9d09-199fff0e4dc9.png)

* Date values are in millis which won’t be very helpful to users 
![screen shot 2017-05-11 at 11 37 16 am](https://cloud.githubusercontent.com/assets/6239176/25961592/cde7df30-3648-11e7-8688-93ba4acf4e94.png)

* How should a user enter value for a geo field? Once you get to the value input there’s no autocomplete and the loading bar just starts spinning (but I don’t see any errors in the console). It also starts spinning if you edit a geo filter created from a tile map.
* Field autocomplete should not suggest unindexed fields
* [Dashboard] By default a filter will affect all panels on a dashboard, but splitting up the autocomplete field lists by index pattern implies (to me) that the filter will only affect the version of the field that you choose. I like that you can tell where a field is getting pulled from, but I wonder if one big list might be better for this reason. It also makes the autocomplete slightly less useful, since it’s only providing suggestions from one index pattern
![screen shot 2017-05-11 at 12 36 07 pm](https://cloud.githubusercontent.com/assets/6239176/25961614/ddae2e4c-3648-11e7-8c54-69751f9d6979.png)

* HTML in values isn't getting escaped
![screen shot 2017-05-11 at 12 40 21 pm](https://cloud.githubusercontent.com/assets/6239176/25961618/e0ff51e8-3648-11e7-934c-f3ab17afafd8.png)

* Getting an error when typing value of non-string fields

![screen shot 2017-05-11 at 11 34 20 am](https://cloud.githubusercontent.com/assets/6239176/25961544/a8df9afc-3648-11e7-98dc-0e0dfbcadfff.png)

```
TypeError: bValue.startsWith is not a function
    at sortPrefixFirstComparator (https://localhost:5601/pvn/bundles/kibana.bundle.js?v=8467:275341:62)
    at Array.sort (native)
    at sortPrefixFirst (https://localhost:5601/pvn/bundles/kibana.bundle.js?v=8467:275335:48)
    at fn (eval at compile (https://localhost:5601/pvn/bundles/commons.bundle.js?v=8467:35700:16), <anonymous>:4:1026)
    at Object.regularInterceptedExpression (https://localhost:5601/pvn/bundles/commons.bundle.js?v=8467:36780:22)
    at Scope.$digest (https://localhost:5601/pvn/bundles/commons.bundle.js?v=8467:38196:41)
    at https://localhost:5601/pvn/bundles/commons.bundle.js?v=8467:38406:27
    at completeOutstandingRequest (https://localhost:5601/pvn/bundles/commons.bundle.js?v=8467:27885:11)
    at https://localhost:5601/pvn/bundles/commons.bundle.js?v=8467:28162:8
```

* When switching between is and is not the value gets reset. Do you think it should remain the same?

* Are we getting rid of the "Apply All" actions?

alexfrancoeur commented:
@lukasolson wow - this PR has certainly come a long way while I was out. Looks good! I just took a peek and had a few comments.

 - Matt mentioned this at the end of his comment and I thought out this when looking at the context viewer, are we getting rid of bulk actions here? I'm a bit concerned about removing this one. 
 - With this change, it looks as if filters are now index specific. Is that correct? Whether it's incorrect or not, I know some users have their data formatted in a way that can utilize filters for multiple indices.
 - This is a nice to have and Matt hinted at it earlier - it'd be nice to handle dates a bit better. I don't believe this is a blocker for V1 but it would be nice to utilize some sort of calendar here in which you can customize the timestamp (similar to our absolute timeframe widget)
<img width="525" alt="screen shot 2017-05-17 at 7 36 27 am" src="https://cloud.githubusercontent.com/assets/25181823/26152167/cc913782-3ad3-11e7-905d-e08c02874983.png">

- Will we be keeping duplicate actions in the pill and in the expanded UI? If so, it looks like we're missing invert filter here.
<img width="522" alt="screen shot 2017-05-17 at 7 42 28 am" src="https://cloud.githubusercontent.com/assets/25181823/26152347/962c9f32-3ad4-11e7-9c30-3d0f6be4e1ab.png">

 - Also minor but is the modal window meant to follow the filter that's been added? It seems to be pinned in the left hand corner and looks as if it's related to the first pill when opened
<img width="764" alt="screen shot 2017-05-17 at 7 51 38 am" src="https://cloud.githubusercontent.com/assets/25181823/26152656/cf6a8718-3ad5-11e7-87e0-941a5f751050.png">

 - Is there a reason for having a dropdown menu for a single value field? Maybe we replace with an X or something that indicates modifying a single value?

![may-17-2017 07-56-41](https://cloud.githubusercontent.com/assets/25181823/26152814/9b2c6060-3ad6-11e7-8a39-be1a5c4242b8.gif)

I'll keep playing around with this and let you know if I come across anything else.

lukasolson commented:
> You can edit the query DSL when editing a filter, but not when creating one, which seems odd

@Bargs If you create a new filter by editing the DSL instead of selecting a field, then which index pattern would we save along with the filter (`filter.meta.index`)? Just pick the first one in the list?

Bargs commented:
I think it depends on the app. In Discover and Visualize it should probably be the selected index pattern. Dashboard is more complicated. With the query language prototype, I just created filters without an index property. I had to update a couple of the mappers, but other than that it just worked. We could also possibly surface the index pattern in the UI.

Bargs commented on src/core_plugins/kibana/server/routes/api/suggestions/register_value_suggestions.js:
```diff
@@ -0,0 +1,31 @@
+import handleESError from '../../../lib/handle_es_error';
+
+export function registerValueSuggestions(server) {
+  server.route({
```
Some [API tests](https://github.com/elastic/kibana/tree/master/test/api_integration) would be good for this route.

Bargs commented on src/core_plugins/kibana/public/dashboard/panel/panel_utils.js:
```diff
@@ -63,4 +66,18 @@ export class PanelUtils {
+        });
+    })
+      .then(indexPatterns => _.uniq(indexPatterns));
+  }
```
There's no way to get the index pattern from the panels without reloading them?

lukasolson replied:
@Stacey-Gammon Could you take a look at this portion? Is there a better way to get a unique list of index patterns used by the panels in a dashboard?

Stacey-Gammon replied:
You could do something like the following to avoid reloading all the saved objects:

https://github.com/Stacey-Gammon/kibana/commit/bc6390f797287461a8b029bc7670a23ecc8b15a2

lukasolson commented on src/ui/public/filter_editor/filter_editor.html:
```diff
@@ -0,0 +1,159 @@
+          ng-show="filterEditor.operator"
+          field="filterEditor.field"
+          operator="filterEditor.operator"
+          params="filterEditor.params"
```
Well, the main reason was that when you update the field and operator, other things need to happen (like the focus changing, clearing params), but this doesn't need to happen when you update the params.

Bargs commented:
Looked through all the updates, code looks good. I noticed one visual issue, in Timeseries visual builder there's an extra blue row showing up near where the filter bar was:

![screen shot 2017-05-31 at 11 45 46 am](https://cloud.githubusercontent.com/assets/6239176/26643871/ca51297c-4600-11e7-9c9d-ca3786d6799e.png)

I'm still gonna do some more functional testing today, but that's all the feedback I had for the code.

Stacey-Gammon commented:
Functionality LGTM, just the suggestion about an alternative way to grab the index patterns from dashboard without reloading all the saved objects in each panel.

lukasolson commented:
@cjcenizal Would you mind providing some feedback for the styles in this PR?

cjcenizal commented on src/ui/public/filter_bar/filter_bar.html:
```diff
@@ -51,75 +54,68 @@
+    ng-show="showAddFilterButton()"
+  >
+    <div class="filter-description small">
+      <a ng-click="addFilter()">
```
You can add `kbn-accessible-click` here to make this button keyboard-accessible.

lukasolson commented:
> Confirming a change to the filter requires extensive tabbing until the save button is focused. Have you considered saving when the user presses the enter key in the value input?

Yeah, I thought about this. It would be easy enough to implement when a simple input text box is the param input editor, but when we use the ui-select, it gets more complicated because hitting "enter" is one way to open the drop-down. What do you think would be intuitive here @weltenwort?

lukasolson commented:
@cjcenizal I'd like to avoid making big design changes this late in the stage. I'm happy to look into the tabbing in a future release if what we've got here is unintuitive.

lukasolson commented:
> The query DSL editor isn't filled before the filter has been changed. So you create a filter, but then as you go to the DSL-editor, that filter is not prepopulated yet.
This is expected behavior because when you create a filter, there isn't really any query DSL to go off of, you're creating it from scratch (and hopefully you know what you're doing 😬).

> Clicking Pin/Unpin, Enable/Disable in the advanced editor isn't changing the status in the pill right away.
I've removed the pin/unpin/enable/disable in the drop-down as it seems to cause more confusion than it's worth, and the corresponding actions are still in the pill on hover (for now).

> When the advanced editor is expanded, clicking the options on the pill doesn't change the corresponding settings in the advanced editor. E.g. expand the advanced editor on a pill, click the invert-filter button on the pill -> the filter in the advanced editor doesn't change
This should be fixed now. Keep in mind that if you update the operator drop-down from something like an "is" to an "is not", the change will not be reflected in the actual pill until you hit "Save".

> Inverting a bounding box filter does not change the query-DSL. So in the advanced editor there is no way to tell that a bounding box filter has been inverted.
Unfortunately the "negate" flag isn't stored directly in the query DSL, but in the "meta" property of a filter, so it currently won't show up in the query DSL at all.

lukasolson commented:
@Bargs 

> Seems like input boxes with help links (e.g. date value) should be vertically aligned with the rest of the inputs, with the link underneath

I agree. I think this might be more effort than it's worth prior to release, though, because I'm not sure the styles available in the UI framework currently support something like this. Would it be agreeable to open an issue to track progress on this after this is released?

cjcenizal commented:
> I'd like to avoid making big design changes this late in the stage.

Just so I can understand, is the problem that this seems like too much work to get in before release? Or are you concerned that we may decide it's not an improvement after we implement it, thus delaying release?

weltenwort commented:
> hitting "enter" is one way to open the drop-down

Right, and it is in line with native drop-down behavior. How about Ctrl+Enter instead? That would be consistent with the Console shortcut for executing the snippet.

weltenwort commented on src/ui/public/filter_editor/filter_editor.less:
```diff
@@ -0,0 +1,27 @@
+  z-index: 101;
+}
+
+.ui-select-choices--autoWidth {
```
That looks a lot better :+1: I'm wondering if it would be worth breaking with the ui-select naming scheme and go with our camel-case scheme (`uiSelectChoice`) or if it's better to break with our convention. @cjcenizal, any opinion?

cjcenizal replied:
For styles inside of the Kibana src, I try to retain consistency with the existing styles, and only use our conventions if it's a new less file and all the classes within it are new. Since this is a new file, I'd say let's use our convention here and go with camel case.

Bargs commented:
> Would it be agreeable to open an issue to track progress on this after this is released?

Sounds fine if it's not a quick fix.

lukasolson commented:
> Just so I can understand, is the problem that this seems like too much work to get in before release? Or are you concerned that we may decide it's not an improvement after we implement it, thus delaying release?

My biggest concern is that this feature has been through the design cycle several times already, and we've worked towards the current design for awhile, and I'm afraid if we re-enter that cycle it could put the feature at risk. That being said, if we can get this merged and then iterate on the design (before Tuesday), I'm not opposed to that. What do you think?

Bargs commented:
I'm also concerned with making design changes at this point (keep in mind, this needs to go into 5.5). Even if the changes seem simple, they introduce risk and additional code that will need to be reviewed. Those reviews would very likely prompt even more cycles of iteration. 

TBH I think this kind of feedback needs to come earlier in the process. As Lukas said we've gone through many design iterations already.

cjcenizal commented:
@Bargs @lukasolson Yeah, I should have provided feedback earlier. That's on me for not joining your sync meetings and actively participating -- sorry!

If we can address my concerns about content becoming clipped and unusable, then I'm OK with merging this and refining the UX in separate PRs.

cjcenizal commented:
> Right, and it is in line with native drop-down behavior. How about Ctrl+Enter instead? That would be consistent with the Console shortcut for executing the snippet.

@weltenwort According to the [ARIA guidelines](https://www.w3.org/TR/wai-aria-practices-1.1/) we should actually avoid this kind of keyboard shortcut.

> It is essential to avoid conflicts with keys that perform system level functions, such as appplication and window management and display and sound control. In general, this can be achieved by refraining from the following types of assignments.
>
> * Any modifier keys + any of Tab, Enter, Space, or Escape.

I suggest we use Enter to interact with the ui-select if it has focus, but if you're focused on another part of the form that's not a ui-select, then Enter submits the form, i.e. adds/updates the filter.

lukasolson commented:
@cjcenizal 

> Maybe we can make the dropdown a bit wider? The select components tend to get cut off:

I've made it a bit wider... Let me know if you think it should be even wider.

> Also, the select component suggestions tended to flow off the top of the screen for me:

Darn, this has to do with me overriding the styles to make the drop-down a bit taller. For now I've removed the custom styling I introduced so it doesn't get cut off of the top of the screen. We'll have to look into other ways to accomplish making the drop-down taller (so you can see more options).

Could you take one more look? Thanks!

cjcenizal commented on src/ui/public/filter_editor/filter_editor.html:
```diff
@@ -0,0 +1,132 @@
+    </div>
+
+    <div
+      class="kuiFieldGroup kuiVerticalRhythm"
```
If we add `kuiFieldGroup--alignTop` here, then each field will align to the top even if one of them is taller than the others.

![image](https://cloud.githubusercontent.com/assets/1238659/26739723/d2909338-4787-11e7-8b36-dc9df067bc48.png)

cjcenizal commented on src/ui/public/filter_editor/filter_editor.html:
```diff
@@ -0,0 +1,132 @@
+        ></filter-operator-select>
+      </div>
+
+      <div class="kuiFieldGroupSection kuiFieldGroupSection--wide kuiColumn kuiColumn--4">
```
Can we remove the kuiColumn and kuiColumn--4 classes from all of these kuiFieldGroupSection elements? And then create a new class called `filterEditor__wideField` and apply it to this last section?

The visual result will be that this last section will get a little extra room, and the others won't take up more than they need to:

![image](https://cloud.githubusercontent.com/assets/1238659/26739767/fcb6a314-4787-11e7-912b-0b47d9db91d7.png)

The new CSS would be:

```less
.filterEditor__wideField {
  overflow: hidden;
}
```

lukasolson replied:
I tried this out but it looks like the suggestions don't get shown since the overflow is hidden... Did I do something wrong?

cjcenizal replied:
Sorry, I missed that originally! Instead of `overflow: hidden` ,`min-width: 0` will work.

```
.filterEditor__wideField {
  min-width: 0;
}
```

lukasolson replied:
Cool, looks good! What do you think?

cjcenizal replied:
I think it looks good!

alexfrancoeur commented:
@lukasolson I had a quick sync with @Bargs today on two items. Neither of which are blockers for this PR but I thought It'd be worth commenting on.

The first, is that I find it a bit of an awkward experience presenting a hover of a hand for an input box if you cannot click into it for a list of fields. This is a minor styling thing but makes me feel as if I'm doing something wrong. If the input box is not analyzed, it'd be good to remove the hand on hover and treat it as a text input.

![jun-02-2017 15-57-46](https://cloud.githubusercontent.com/assets/25181823/26742600/65008be6-47ac-11e7-922a-4ba04380d598.gif)

When speaking with Matt I had also mentioned that I imagined users wanting a "contains" and "does not contain" capability soon after implementing this. He informed me that this is possible with using "is" or "is not" if the field has not been analyzed. Matt had suggested that in a future PR, we do not separate the normal fields from the analyzed fields and add "contains" to the menu. If a user chooses "contains" we simply filter on the field that has not been analyzed. If they choose "is" we filter on the analyzed field and provide a drop down. Again, not expected for this PR but I wanted to +1 the concept and maybe link to this comment in a future PR 😄

spalger commented:
![dance](https://cloud.githubusercontent.com/assets/1329312/26749103/db63deca-47b9-11e7-9c6a-d190c207e1d5.gif)
